### PR TITLE
[Drop Keyspace] Oracle wrap up

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
Lets you delete oracle namespaces! Does not clean up the mapping table nor the metadata table.

**After this PR**:
Can clean up oracle namespaces! No one can really instantiate this easily, the factories are coming later

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:
Should I be optionally deleting the internal metadata and mapping table? e.g add a config
I'm using all tables rather than user_tables, for no major reason other than flexibility of filtering
(addendum: for tests_ -> sqlite oracle adapter lets me test what happens in the presence of other users. Arguably none of that would be necessary _if_ we trust user_tables)
**Is documentation needed?**:
No
## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No
**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
Yes, by deleting it. Yes
**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes
**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No
**Does this PR need a schema migration?**
No
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
that oracle doesn't change up their internal tables. Deleting data will fail if they do
**What was existing testing like? What have you done to improve it?**:
Lots of tests added
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
None
**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A
## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
Data deleted
**Has the safety of all log arguments been decided correctly?**:
No logs TODO will add some when I'm back
**Will this change significantly affect our spending on metrics or logs?**:
No
**How would I tell that this PR does not work in production? (monitors, etc.)**:
Data not deleted (will add logs)
**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Don't use the feature
**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:
N/A
## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
I delete sequentially, so no
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
Yes
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
I could add some logic to backoff when deleting a bunch of data - but deletes are generally cheap.
## Development Process
**Where should we start reviewing?**:
OracleNamespaceDeleter
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
I split it up as best I could, but tests mannn
**Please tag any other people who should be aware of this PR**:
@jeremyk-91

### CODE REVIEW

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        return OracleNamespaceDeleterParameters.builder()
+                .tablePrefix(ddlConfig.tablePrefix())
+                .overflowTablePrefix(ddlConfig.overflowTablePrefix())
+                .userId(TEST_USER)
+                .oracleDdlTableFactory(ddlTableFactory)
+                .connectionSupplier(connectionSupplier)
+                .tableNameGetter(tableNameGetter);
+    }
+
+    private class UnstableConnectionSupplier extends ConnectionSupplier {
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java]: Would happily take an alternative way of doing this, but my goal is to throw in the middle of deleting some tables - I guess I could mock out the factory function and have that throw? But that got quite messy with extracting out the dependency tree such that we continue deleting after the fact

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+        assertThat(listAllPhysicalTableNames()).hasSize(20);
+        assertThatThrownBy(namespaceDeleter::deleteAllDataFromNamespace);
+        assertThat(listAllPhysicalTableNames()).hasSizeLessThan(20);
+    }
+
+    @Test
+    public void deleteAllDataFromNamespaceIsRetryable() {
+        NamespaceDeleter namespaceDeleter =
+                new OracleNamespaceDeleter(createNamespaceDeleterParameters(new UnstableConnectionSupplier(25))
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java]: 25 is a magic number :/ see the comment in unstable connection supplier.

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    private OracleNamespaceDeleterParameters createNamespaceDeleterParametersWithUnstableFactory() {
+        OracleNamespaceDeleterParameters parameters =
+                createDefaultNamespaceDeleterParameters().build();
+        return OracleNamespaceDeleterParameters.builder()
+                .from(parameters)
+                .oracleDdlTableFactory(new UnstableOracleDdlFactory(parameters.oracleDdlTableFactory()))
+                .build();
+    }
+
+    private static class UnstableOracleDdlFactory implements Function<TableReference, OracleDdlTable> {
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:345]: I had an unstable connection supplier but that involved more magic numbers, so this seemed a better way to induce failures in the middle of deleting a bunch of tables whilst ensuring we can still test retries.

happy to take an alternative!

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+                .connectionSupplier(connectionSupplier)
+                .tableNameGetter(oracleTableNameGetter)
+                .oracleDdlTableFactory(ddlTableFactory)
+                .userId(connectionConfig.getDbLogin())
+                .build());
+    }
+
+    @After
+    public void after() {
+        namespaceDeleter.deleteAllDataFromNamespace();
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:148]: I do use the namespace deleter itself to clean up, so obviously if it doesn't work, all tests will fail. However, it is the easiest way of cleaning up the state enough for the tests. Another alternative I considered was just forcing the container shuts down and restarts, but that's nontrivial and expensive. Additionally. I could use different users, but then I need to setup the docker container to allow for more users...

Comment (mdaudali) [reply atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:148]: Which does highlight one flaw of these tests... I don't test across different users.

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+
+        assertThat(namespaceDeleter.isNamespaceDeletedSuccessfully()).isTrue();
+    }
+
+    private void createTwoLogicalTablesWithDefaultPrefix() {
+        AllTableDetailsSnapshot kvsTablesSnapshot = getTableDetailsForDefaultPrefixes();
+        assertSnapshotIsNotEmpty(kvsTablesSnapshot);
+        createTableAndOverflowTable(TABLE_NAME_ONE);
+        createTableAndOverflowTable(TABLE_NAME_TWO);
+        assertSnapshotContainsNewTableReferences(
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:231]: Just to check that we're actually creating tables!

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    }
+
+    private static void assertSnapshotIsNotEmpty(AllTableDetailsSnapshot snapshot) {
+        assertThat(snapshot.prefixedTableNames()).isNotEmpty();
+        assertThat(snapshot.physicalTableNames()).isNotEmpty();
+        assertThat(snapshot.tableReferences()).isNotEmpty();
+    }
+
+    @Value.Immutable
+    interface AllTableDetailsSnapshot {
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:353]: Helper immutable to make testing that we're cleaning up _everything_ required easier! kinda related to the TableDetails in sqliteoracletable which represents info for a single table, but the issue here is we can't trivially construct a mapping each prefixed table to a physical table to a table reference (without pulling in the table mappers and stuff), but nor do I really need that granularity for these tests.

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/DbKvsOracleTestSuite.java [hunk 7]
```diff
@@ -108,12 +110,16 @@ public static ConnectionSupplier getConnectionSupplier() {
...
 
     public static ConnectionSupplier getConnectionSupplier(KeyValueService kvs) {
-        ConnectionManagerAwareDbKvs castKvs = (ConnectionManagerAwareDbKvs) kvs;
         ReentrantManagedConnectionSupplier connSupplier =
-                new ReentrantManagedConnectionSupplier(castKvs.getConnectionManager());
+                new ReentrantManagedConnectionSupplier(getConnectionManager(kvs));
         return new ConnectionSupplier(getSimpleTimedSqlConnectionSupplier(connSupplier));
     }
 
+    public static ConnectionManager getConnectionManager(KeyValueService kvs) {
```
Comment (mdaudali) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/DbKvsOracleTestSuite.java:118]: https://github.com/palantir/atlasdb/pull/6264/files#diff-e09fb7e764291e48f1849d52da8faf3316471d2f5132e79aa64ce52c570cebedR94

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+                // tests that case.
+                createTable("ab_", TABLE_NAME_1, TEST_USER));
+
+        assertThatTableDetailsMatchPersistedData(Sets.union(tablesToDelete, tablesThatShouldNotBeDeleted));
+
+        namespaceDeleter.deleteAllDataFromNamespace();
+        assertThatTableDetailsMatchPersistedData(tablesThatShouldNotBeDeleted);
+    }
+
+    @Test
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:108]: Maybe silly: I'd probably just add one with no tables at all, just to make sure it works

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+        assertThat(listAllPhysicalTableNames()).hasSizeBetween(1, 19);
+
+        namespaceDeleter.deleteAllDataFromNamespace();
+        assertThat(listAllPhysicalTableNames()).isEmpty();
+    }
+
+    @Test
+    public void isNamespaceDeletedSuccessfullyDoesNotExecuteArbitrarySqlOnOwner() {
+        NamespaceDeleter namespaceDeleter = new OracleNamespaceDeleter(createDefaultNamespaceDeleterParameters()
+                .userId("1'; CREATE TABLE mwahahaha (evil VARCHAR(128) PRIMARY KEY); --")
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:178]: ❤️

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+            createTablesWithDefaultPrefixes(TABLE_NAME_1 + i);
+        }
+
+        assertThat(listAllPhysicalTableNames()).hasSize(20);
+        assertThatThrownBy(namespaceDeleter::deleteAllDataFromNamespace);
+        assertThat(listAllPhysicalTableNames()).hasSizeBetween(1, 19);
+
+        namespaceDeleter.deleteAllDataFromNamespace();
+        assertThat(listAllPhysicalTableNames()).isEmpty();
+    }
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:173]: This makes sense - unsure if we need the previous test

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+        return new OracleNamespaceDeleter(createNamespaceDeleterParameters(
+                        getDefaultDdlConfig().useTableMapping(false).build())
+                .build());
+    }
+
+    private ImmutableOracleDdlConfig.Builder getDefaultDdlConfig() {
+        return ImmutableOracleDdlConfig.builder()
+                .tablePrefix(TABLE_PREFIX)
+                .overflowTablePrefix(OVERFLOW_TABLE_PREFIX)
+                .overflowMigrationState(OverflowMigrationState.UNSTARTED)
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:410]: nit: use `FINISHED`. It doesn't really matter but in practice is what we run with nowadays in prod everywhere

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    private OracleNamespaceDeleterParameters createNamespaceDeleterParametersWithUnstableFactory() {
+        OracleNamespaceDeleterParameters parameters =
+                createDefaultNamespaceDeleterParameters().build();
+        return OracleNamespaceDeleterParameters.builder()
+                .from(parameters)
+                .oracleDdlTableFactory(new UnstableOracleDdlFactory(parameters.oracleDdlTableFactory()))
+                .build();
+    }
+
+    private static final class UnstableOracleDdlFactory implements Function<TableReference, OracleDdlTable> {
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:449]: nit: docs, or rename to maybe `OnceFailingOracleDdlFactory`? Main thing is I perceive `Unstable` to fail infinitely often as infinitely many requests are being made, and while that's _technically_ true, 1 in 2^32 is not really unstable.

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+    private static final String TABLE_NAME_3 = "world";
+    private static final String TEST_USER = "testuser";
+    private static final String TEST_USER_2 = "testuser2";
+
+    private ExecutorService executorService;
+    private SqliteOracleAdapter sqliteOracleAdapter;
+
+    @Before
+    public void before() throws IOException {
+        executorService = MoreExecutors.newDirectExecutorService();
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterTests.java:76]: Is there any reason we need this to be shared across invocations in the same test method? If not, it seems that you can just get away with passing this in wherever it's used

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+
+    public OracleNamespaceDeleterIntegrationTest() {
+        super(TRM, TRM);
+    }
+
+    @Before
+    public void before() {
+        dbKeyValueServiceConfig = DbKvsOracleTestSuite.getKvsConfig();
+        oracleDdlConfig = (OracleDdlConfig) dbKeyValueServiceConfig.ddl();
+        executorService = MoreExecutors.newDirectExecutorService();
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:101]: Is this intentionally shared across all of the deleters?

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:175]: Yes. How hard would it be to add these? I'd recommend just proceeding with this PR and doing them subsequently, given this PR's size.

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+        return builder.build();
+    }
+
+    private static String withEscapedUnderscores(String s) {
+        return s.replace("_", "\\_");
+    }
+
+    private static String withWildcardSuffix(String s) {
+        return s + "%";
+    }
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:432]: can we have a more concrete name for `s`?

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+
+    private static void assertSnapshotContainsNewTableReferences(
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:434]: maybe not actionable: I don't really like this on the basis of it checks more than what it says (it checks that the snapshot is a 'normal' looking snapshot with the given number of TRs). Though I don't really have a better name

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,483 @@
...
+        assertSnapshotIsEmptyExcludingTimestampTable(getTableDetailsForDefaultPrefixes());
+    }
+
+    @Test
+    public void deleteAllDataFromNamespaceDoesNotThrowWhenNoTablesToDelete() {
+        // Oracle throws on an empty IN () clause, whereas sqlite does not.
+        namespaceDeleter.deleteAllDataFromNamespace();
+        assertSnapshotIsEmptyExcludingTimestampTable(getTableDetailsForDefaultPrefixes());
+
+        assertThatNoException().isThrownBy(namespaceDeleter::deleteAllDataFromNamespace);
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:215]: nit: Is fine, though convention is normally `assertThatCode(namespaceDeleter::deleteAllDataFromNamespace).doesNotThrowAnyException();`

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+        assertThat(namespaceDeleter.isNamespaceDeletedSuccessfully()).isTrue();
+    }
+
+    @Test
+    public void isNamespaceDeletedSuccessfullyReturnsTrueEvenIfOtherPrefixExists() {
+        namespaceDeleter.deleteAllDataFromNamespace();
+        assertThat(namespaceDeleter.isNamespaceDeletedSuccessfully()).isTrue();
+        createTableAndOverflowTableWithNonDefaultPrefix(TABLE_NAME_ONE);
+
+        assertThat(namespaceDeleter.isNamespaceDeletedSuccessfully()).isTrue();
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:286]: nit: This is probably clearer if we just assert the non-default prefix?

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,483 @@
...
+    @Test
+    public void isNamespaceDeletedSuccessfullyReturnsTrueIfUnmappedTablesExistForMappedUser() {
+        namespaceDeleter.deleteAllDataFromNamespace();
+        createTableAndOverflowTableWithDefaultPrefixNoMapping(TABLE_NAME_ONE);
+
+        assertThat(namespaceDeleter.isNamespaceDeletedSuccessfully()).isTrue();
+    }
+
+    @Test
+    public void isNamespaceDeletedSuccessfullyReturnsTrueIfMappedTablesExistForUnMappedUser() {
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:298]: nit: you're using `unmapped` as 1 word

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+
+    private void assertSnapshotIsEmptyExcludingTimestampTable(AllTableDetailsSnapshot snapshot) {
+        assertThat(snapshot.prefixedTableNames())
+                .hasSameSizeAs(snapshot.tableReferences())
+                .isEmpty();
+        if (snapshot.physicalTableNames().size() == 1) {
+            assertThat(snapshot.physicalTableNames().stream()
+                            .map(s -> s.toLowerCase(Locale.ROOT))
+                            .collect(Collectors.toSet()))
+                    .containsOnly(timestampTableName);
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:449]: nit: The stream's a bit unnecessary: you can just do `            assertThat(Iterables.getOnlyElement(snapshot.physicalTableNames()).toLowerCase(Locale.ROOT)).isEqualTo(timestampTableName);`.

**Review by jeremyk-91 — COMMENTED** [review]
Makes sense. Coverage is generally good: I'd have wanted "different Oracle user", though given the size plus potential difficulty I'd suggest you timebox it and also do it in a separate PR, we should get this in first. I also have a question on `all_tables` vs `user_tables`

Concerns:

>Should I be optionally deleting the internal metadata and mapping table? e.g add a config

Maybe, but definitely not right now - let's get this merged first.

> I'm using all tables rather than user_tables, for no major reason other than flexibility of filtering

What do you mean by flexibility? `user_tables` would seem more appropriate, so would prefer that if it's not too hard to write.

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/cleaner/OracleNamespaceDeleter.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+    private Set<TableReference> getAllTableNamesToDrop() {
+        Set<String> nonOverflowTables = getAllNonOverflowTables();
+        Set<String> overflowTables = getAllOverflowTables();
+
+        try {
+            return Sets.union(
+                    tableNameGetter.getTableReferencesFromShortTableNames(connectionSupplier, nonOverflowTables),
+                    tableNameGetter.getTableReferencesFromShortOverflowTableNames(connectionSupplier, overflowTables));
+        } catch (TableMappingNotFoundException e) {
+            throw Throwables.rewrapAndThrowUncheckedException(e);
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/cleaner/OracleNamespaceDeleter.java]: yeah, the idea is that if `e` is an RE we don't want an RE of an RE of something checked I think

**Comment by mdaudali** [comment]
(Discussed offline for all_tables -> user_tables - we're keeping all_tables for now to aid in testing. May revisit if we can add integration tests that create new users)

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/cleaner/OracleNamespaceDeleter.java [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+import com.palantir.atlasdb.namespacedeleter.NamespaceDeleter;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.immutables.value.Value;
+
+/**
+ * Drops all Oracle Tables associated with the table and overflow table prefixes provided in the config.
+ * This does _not_ clean up the AtlasDB metadata, _timestamp, and name mapping tables, as these can be shared
+ * across multiple prefixes.
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/cleaner/OracleNamespaceDeleter.java:35]: Fair enough. I didn't find anything else, though integration tests should help give us a bit more confidence :)

File: atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java (hunk 18 above)
Comment (jeremyk-91) [reply atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleNamespaceDeleterIntegrationTest.java:284]: Ah. Fair enough, reading this comment with hindsight I'm not sure what it was referring to

**Review by jeremyk-91 — APPROVED** [review]
👍

**Pull request merged**